<div dir="rtl" style="text-align:right;">

# فصل ۱۲: کار با فایل‌ها در پایتون
</div>


<div dir="rtl" style="text-align:right;">

## ۱۰.۱ مقدمه

در فصل‌های قبل، داده‌ها را در متغیرها ذخیره یا از کاربر دریافت می‌کردیم. اما این داده‌ها با پایان برنامه از بین می‌روند. برای ذخیره دائمی اطلاعات، نیاز به کار با **فایل‌ها** داریم.

فایل‌ها می‌توانند حاوی متن (مثل فایل‌های `.txt` یا `.csv`) یا داده‌های باینری (مثل عکس‌ها و ویدئوها) باشند. پایتون ابزارهای قدرتمندی برای خواندن و نوشتن در فایل‌ها فراهم می‌کند.

در این فصل یاد می‌گیریم:
- باز کردن و بستن فایل‌ها
- خواندن محتوای فایل
- نوشتن در فایل‌ها
- کار با حالت‌های مختلف فایل
- استفاده از `with open()` برای مدیریت خودکار

---

## ۱۰.۲ باز کردن و بستن فایل

برای کار با یک فایل، ابتدا باید آن را **باز** کنید. این کار با تابع `open()` انجام می‌شود.

# ساختار کلی
</div>


In [ ]:
file_object = open(filename, mode)

<div dir="rtl" style="text-align:right;">

- `filename`: نام فایل (مثلاً `"data.txt"`)
- `mode`: حالتی که فایل را باز می‌کنید (مثلاً `'r'` برای خواندن)

پس از اتمام کار، **حتماً** باید فایل را **ببندید** تا منابع آزاد شوند و داده‌ها ذخیره شوند.
</div>


In [ ]:
file = open("my_notes.txt", "w")  # باز کردن برای نوشتن
    # ... اینجا کارهایی روی فایل انجام می‌دهیم ...
file.close()  # بستن فایل

<div dir="rtl" style="text-align:right;">

---

## ۱۰.۳ حالت‌های باز کردن فایل (Modes)

| حالت | شرح |
|---|---|
| `'r'` | **Read (خواندن)**: پیش‌فرض. فایل را برای خواندن باز می‌کند. خطا می‌دهد اگر فایل وجود نداشته باشد. |
| `'w'` | **Write (نوشتن)**: فایل را برای نوشتن باز می‌کند. اگر فایل وجود داشته باشد، محتوای قبلی پاک می‌شود. اگر وجود نداشته باشد، ساخته می‌شود. |
| `'a'` | **Append (اضافه کردن)**: محتوا به انتهای فایل اضافه می‌شود. اگر فایل وجود نداشته باشد، ساخته می‌شود. |
| `'x'` | **Create (ساختن)**: فایل را برای ساختن باز می‌کند. اگر فایل وجود داشته باشد، خطا می‌دهد. |
| `'b'` | **Binary (باینری)**: برای کار با فایل‌های غیرمتنی. با حالت‌های دیگر ترکیب می‌شود (مثلاً `'rb'` یا `'wb'`). |
| `'+'` | **Update (خواندن و نوشتن)**: با حالت‌های دیگر ترکیب می‌شود (مثلاً `'r+'` یا `'w+'`). |

---

## ۱۰.۴ خواندن از فایل

فرض کنید فایلی به نام `sample.txt` با محتوای زیر داریم:

> در مسیر پروژه یک فایل به نام `sample.txt` بسازید و محتوای زیر را در آن قرار دهید.

```txt
Line 1: Hello World
Line 2: This is a sample file.
Line 3: For reading demonstration.
```
### ۱. خواندن کل محتوا با `read()`
</div>


In [1]:
file = open("data/sample.txt", "r")
content = file.read()
print(content)
file.close()

Line 1: Hello World
Line 2: This is a sample file.
Line 3: For reading demonstration.


<div dir="rtl" style="text-align:right;">

خروجی:

Line 1: Hello World
Line 2: This is a sample file.
Line 3: For reading demonstration.

### ۲. خواندن خط به خط با `readline()`
</div>


In [2]:
file = open("data/sample.txt", "r")
line1 = file.readline()
line2 = file.readline()
print(f"خط اول: {line1.strip()}")  # strip() برای حذف \n آخر خط
print(f"خط دوم: {line2.strip()}")
file.close()

خط اول: Line 1: Hello World
خط دوم: Line 2: This is a sample file.


<div dir="rtl" style="text-align:right;">

خروجی:

خط اول: Line 1: Hello World
خط دوم: Line 2: This is a sample file.

### ۳. خواندن همه خطوط در لیست با `readlines()`
</div>


In [3]:
file = open("data/sample.txt", "r")
lines = file.readlines()
print(lines)
file.close()

['Line 1: Hello World\n', 'Line 2: This is a sample file.\n', 'Line 3: For reading demonstration.']


<div dir="rtl" style="text-align:right;">

خروجی:

['Line 1: Hello World\n', 'Line 2: This is a sample file.\n', 'Line 3: For reading demonstration.']

### پیمایش فایل با حلقه `for` (روش پیشنهادی)

این روش حافظه کمتری مصرف می‌کند چون کل فایل را یکجا بارگذاری نمی‌کند.
</div>


In [4]:
file = open("data/sample.txt", "r")
print("محتوای فایل:")
for line in file:
    print(f"  -> {line.strip()}")
file.close()

محتوای فایل:
  -> Line 1: Hello World
  -> Line 2: This is a sample file.
  -> Line 3: For reading demonstration.


<div dir="rtl" style="text-align:right;">

خروجی:

محتوای فایل:
  -> Line 1: Hello World
  -> Line 2: This is a sample file.
  -> Line 3: For reading demonstration.

---

## ۱۰.۵ نوشتن در فایل

### حالت `'w'` (Write) — بازنویسی
</div>


<div dir="rtl" style="text-align:right;">

# اگر فایل موجود باشد، پاک شده و از نو نوشته می‌شود
</div>


In [5]:
with open("output.txt", "w") as f:
    f.write("This is the first line.\n")
    f.write("This overwrites the previous content.\n")

<div dir="rtl" style="text-align:right;">

### حالت `'a'` (Append) — اضافه کردن
</div>


<div dir="rtl" style="text-align:right;">

# محتوا به انتهای فایل اضافه می‌شود
</div>


In [15]:
with open("output.txt", "a") as f:
    f.write("This line is appended.\n")

<div dir="rtl" style="text-align:right;">

### نوشتن چند خط با `writelines()`

`writelines()` یک لیست از رشته‌ها را می‌گیرد و در فایل می‌نویسد.

> **توجه:** `writelines()` خودش `\n` اضافه نمی‌کند. باید در هر رشته آن را بگنجانید.
</div>


In [6]:
lines_to_write = ["First line\n", "Second line\n", "Third line\n"]
with open("output_lines.txt", "w") as f:
    f.writelines(lines_to_write)

<div dir="rtl" style="text-align:right;">

---

## ۱۰.۶ استفاده از `with open()` — بهترین روش

`with open(...) as ...:` تضمین می‌کند که فایل **حتماً** پس از اتمام بلوک کد بسته شود، حتی اگر خطا رخ دهد.
</div>


In [7]:
try:
    with open("data/sample.txt", "r") as infile:
        content = infile.read()
        print("فایل خوانده شد.")

    with open("output.txt", "a") as outfile:
        outfile.write("Appended using 'with'.\n")
        print("به فایل اضافه شد.")

except FileNotFoundError:
    print("خطا: فایل sample.txt یافت نشد.")
except Exception as e:
    print(f"خطایی رخ داد: {e}")


فایل خوانده شد.
به فایل اضافه شد.


<div dir="rtl" style="text-align:right;">

# نیازی به file.close() نیست

---

## ۱۰.۷ کار با فایل‌های متنی و باینری

- **فایل‌های متنی (`'r'`, `'w'`, `'a'`):** پایتون به طور خودکار کاراکترهای پایان خط (`\n`) را مدیریت می‌کند.
- **فایل‌های باینری (`'rb'`, `'wb'`, `'ab'`):** برای داده‌هایی که متن نیستند (عکس، فایل فشرده، داده‌های خام). `read()` بایت برمی‌گرداند و `write()` بایت می‌گیرد.
</div>


<div dir="rtl" style="text-align:right;">

# خواندن تصویر (باینری)
</div>


In [8]:
try:
    with open("data/image.jpeg", "rb") as img_file:
        image_data = img_file.read()
        print(f"اندازه تصویر: {len(image_data)} بایت")
except FileNotFoundError:
    print("فایل image.jpg پیدا نشد.")

اندازه تصویر: 84152 بایت


<div dir="rtl" style="text-align:right;">

# نوشتن داده باینری
</div>


In [9]:
binary_data = b'\x01\x02\x03\x04\x05'
with open("data.bin", "wb") as bin_file:
    bin_file.write(binary_data)
    print("فایل باینری data.bin نوشته شد.")

فایل باینری data.bin نوشته شد.


<div dir="rtl" style="text-align:right;">

---

## ۱۰.۸ تعیین انکدینگ (Encoding)

وقتی با فایل‌های متنی کار می‌کنید، پایتون از انکدینگ پیش‌فرض سیستم عامل استفاده می‌کند. برای اطمینان، همیشه انکدینگ را صریح مشخص کنید.
</div>


<div dir="rtl" style="text-align:right;">

# خواندن با انکدینگ مشخص
</div>


In [10]:
with open("data/persian_text.txt", "r", encoding="utf-8") as f:
    content = f.read()
    print(content)

این یک متن تست فارسی است


<div dir="rtl" style="text-align:right;">

# نوشتن با انکدینگ مشخص
</div>


In [11]:
with open("unicode_example.txt", "w", encoding="utf-8") as f:
    f.write("سلام دنیا\n")
    f.write("Hello World\n")

<div dir="rtl" style="text-align:right;">

> **توصیه:** همیشه `encoding="utf-8"` را در پروژه‌های خود مشخص کنید تا از مشکلات ناسازگاری انکدینگ جلوگیری شود.

---

## ۱۰.۹ مثال کاربردی: خواندن و پردازش فایل CSV

فرض کنید فایلی به نام `data.csv` داریم:
</div>


<div dir="rtl" style="text-align:right;">

</div>


<div dir="rtl" style="text-align:right;">

در این مثال، می‌خواهیم یک فایل CSV واقعی را پردازش کنیم. در پوشه `data` فایلی به نام `Data1.csv` وجود دارد که حاوی داده‌های زمین‌شناسی است.

این فایل شامل ستون‌های زیر است:
- `rock_name`: نام سنگ
- `long`: طول جغرافیایی
- `lat`: عرض جغرافیایی
- `SiO2n`, `TiO2n`, `Al2O3n`, ...: درصد اکسیدهای شیمیایی

</div>


In [14]:
import csv

rocks_data = []

try:
    with open("data/Data1.csv", "r", encoding="utf-8") as csvfile:
        # خواندن CSV به صورت دیکشنری
        reader = csv.DictReader(csvfile)

        for row in reader:

            # تبدیل مقادیر عددی؛ اگر خالی باشند None قرار می‌دهیم
            row['long'] = float(row['long']) if row['long'] else None
            row['lat'] = float(row['lat']) if row['lat'] else None
            row['SiO2n'] = float(row['SiO2n']) if row['SiO2n'] else None

            rocks_data.append(row)

    print(f"تعداد نمونه‌های خوانده شده: {len(rocks_data)}")

    print("\nنمونه‌ای از داده‌ها:")

    for i, rock in enumerate(rocks_data[:3]):
        print(
            f"{i + 1}. "
            f"Long: {rock['long']} | "
            f"Lat: {rock['lat']} | "
            f"SiO2: {rock['SiO2n']}%"
        )

except FileNotFoundError:
    print("خطا: فایل Data1.csv یافت نشد.")

except Exception as e:
    print(f"خطا در پردازش CSV: {e}")

تعداد نمونه‌های خوانده شده: 4162

نمونه‌ای از داده‌ها:
1. Long: -122.585 | Lat: 46.23 | SiO2: 47.77%
2. Long: -122.5806 | Lat: 46.2436 | SiO2: 47.94%
3. Long: -122.5925 | Lat: 46.2036 | SiO2: 51.5%


<div dir="rtl" style="text-align:right;">

### روش دوم: استفاده از کتابخانه pandas (پیشنهادی)

کتابخانه `pandas` یکی از قدرتمندترین ابزارها برای کار با داده‌های جدولی در پایتون است. این کتابخانه به راحتی فایل‌های CSV را می‌خواند و امکانات زیادی برای تحلیل داده فراهم می‌کند.

> دقت کنید که قبل از اجرای این سلول باید کتابخانه `pandas` رو روی سیستم خودتون نصب کنید

> میتونید از گوگل کولب برای اجرای این نوت بوک ها استفاده کنید و بدون دردسر سلول ها را اجرا کنید
</div>


In [ ]:
import pandas as pd

# خواندن فایل CSV
df = pd.read_csv("data/Data1.csv")

df


,rock_name,long,lat,SiO2n,TiO2n,Al2O3n,FeO*n,MnOn,MgOn,CaOn,Na2On,K2On,P2O5n
0,Basalt,-122.5850,46.2300,47.77,1.33,15.38,9.59,0.17,10.76,11.28,2.69,0.62,0.41
1,Basalt,-122.5806,46.2436,47.94,1.28,15.22,9.60,0.18,10.87,11.39,2.17,0.96,0.38
2,Basalt,-122.5925,46.2036,51.50,1.12,16.04,8.77,0.15,9.49,9.80,2.64,0.33,0.16
3,Basalt,-122.5797,46.2144,45.66,1.60,14.87,9.90,0.17,9.81,13.09,3.09,0.96,0.84
4,Basalt,-122.5839,46.2100,48.24,1.34,15.91,10.16,0.17,10.06,10.78,2.51,0.56,0.26
...,...,...,...,...,...,...,...,...,...,...,...,...,...
4157,Andesite,-121.5653,47.3017,57.65,0.49,18.54,6.88,0.15,3.43,7.86,3.53,1.31,0.15
4158,Dacite,-121.6000,47.2675,69.97,0.49,15.04,3.97,0.08,0.38,2.35,5.12,2.46,0.14
4159,Dacite,-121.7633,47.4092,64.97,0.48,17.18,4.98,0.10,1.77,5.62,3.75,0.97,0.19
4160,Andesite,-121.0000,47.3017,58.04,1.02,19.41,7.35,0.12,2.04,7.15,3.78,0.81,0.28


<div dir="rtl" style="text-align:right;">

### تحلیل داده‌ها با pandas

حالا می‌توانیم عملیات مختلفی روی داده‌ها انجام دهیم:

</div>


In [32]:
# آمار توصیفی داده‌های عددی
print("آمار توصیفی:")
print(df.describe())


آمار توصیفی:
              long          lat        SiO2n        TiO2n       Al2O3n  \
count  4118.000000  4118.000000  4135.000000  4135.000000  4135.000000   
mean   -122.193137    45.541801    58.595127     1.164000    16.801287   
std       0.432682     1.270632     6.917297     0.487531     1.588845   
min    -123.249000    41.656000    45.420000     0.000000     3.230000   
25%    -122.533075    44.652000    53.375000     0.850000    15.870000   
50%    -122.254250    46.099000    56.990000     1.150000    16.830000   
75%    -121.881000    46.375800    62.580000     1.440000    17.760000   
max    -120.531700    48.358300    83.730000     3.640000    25.040000   

             FeO*n         MnOn         MgOn         CaOn        Na2On  \
count  4135.000000  3980.000000  4135.000000  4135.000000  4134.000000   
mean      7.430377     0.141910     3.679231     7.099359     3.556638   
std       2.580065     0.142415     2.112921     2.741486     0.752476   
min       0.110000     0

In [31]:
# گروه‌بندی بر اساس نوع سنگ و محاسبه میانگین
print("\nمیانگین SiO2 برای هر نوع سنگ:")
rock_groups = df.groupby('rock_name')['SiO2n'].mean()
print(rock_groups)



میانگین SiO2 برای هر نوع سنگ:
rock_name
Altered andesite                  61.290000
Andesite                          59.447938
Andesite                          60.053333
Andesite  sill                    58.290000
Andesite and microdiorite sill    61.590000
                                    ...    
Rhyolite sill                     73.390000
Rhyolite tuff                     74.511887
Silicic dike                      69.240000
Tonalite                          69.580000
Volcanic breccia                  58.530000
Name: SiO2n, Length: 97, dtype: float64


In [16]:
# فیلتر کردن داده‌ها: سنگ‌هایی با SiO2 بیشتر از 50
high_silica = df[df['SiO2n'] > 50]
print(f"\nتعداد سنگ‌های با SiO2 > 50: {len(high_silica)}")
print(high_silica[['rock_name', 'SiO2n', 'long', 'lat']].head())



تعداد سنگ‌های با SiO2 > 50: 3908
            rock_name  SiO2n      long      lat
2              Basalt  51.50 -122.5925  46.2036
6   Basaltic andesite  52.54 -122.5789  46.2078
8              Basalt  50.59 -122.5731  46.2275
9              Basalt  50.38 -122.6200  46.2433
10             Basalt  50.61 -122.6092  46.1800


In [17]:
# پیدا کردن ماکزیمم و مینیمم
max_SiO2 = df.loc[df['SiO2n'].idxmax()]
min_SiO2 = df.loc[df['SiO2n'].idxmin()]

print(f"بیشترین SiO2: {max_SiO2['SiO2n']}% - نوع سنگ: {max_SiO2['rock_name']}")
print(f"کمترین SiO2: {min_SiO2['SiO2n']}% - نوع سنگ: {min_SiO2['rock_name']}")


بیشترین SiO2: 83.73% - نوع سنگ: Rhyolite
کمترین SiO2: 45.42% - نوع سنگ: Basalt


<div dir="rtl" style="text-align:right;">

این کد با استفاده از ماژول `csv` داده‌ها را می‌خواند و نتایج زیر را نمایش می‌دهد:

تعداد نمونه‌های خوانده شده: 4213

نمونه‌ای از داده‌ها:
1. Basalt - SiO2: 47.77%
2. Basalt - SiO2: 47.94%
3. Basalt - SiO2: 51.50%

</div>


In [18]:
source_file      = "data/sample.txt"
destination_file = "data/sample_copy.txt"

try:
    with open(source_file, "rb") as infile, open(destination_file, "wb") as outfile:
        chunk_size = 4096  # 4KB

        while True:
            chunk = infile.read(chunk_size)
            if not chunk:
                break  # پایان فایل
            outfile.write(chunk)

    print(f"فایل '{source_file}' با موفقیت به '{destination_file}' کپی شد.")

except FileNotFoundError:
    print(f"خطا: فایل منبع '{source_file}' یافت نشد.")
except Exception as e:
    print(f"خطای کپی فایل: {e}")


فایل 'data/sample.txt' با موفقیت به 'data/sample_copy.txt' کپی شد.


<div dir="rtl" style="text-align:right;">

---

## ۱۰.۱۱ نکات مهم

> **نکته ۱ — همیشه از `with open()` استفاده کنید:** این روش تضمین می‌کند فایل حتی در صورت بروز خطا بسته شود.

> **نکته ۲ — حالت `'w'` خطرناک است:** اگر فایل از قبل وجود داشته باشد، تمام محتوای آن بدون هشدار پاک می‌شود. قبل از استفاده مطمئن شوید.

> **نکته ۳ — انکدینگ را فراموش نکنید:** برای فایل‌های حاوی متن فارسی یا کاراکترهای غیر ASCII، همیشه `encoding="utf-8"` را مشخص کنید.

> **نکته ۴ — برای فایل‌های بزرگ از حلقه استفاده کنید:** به جای `read()` که کل فایل را در حافظه بارگذاری می‌کند، با حلقه `for` خط به خط بخوانید.
</div>


<div dir="rtl" style="text-align:right;">

---

## ۱۰.۱۲ تمرین‌های عملی

### تمرین ۱ — ایجاد فایل متنی
برنامه‌ای بنویسید که از کاربر چند خط متن بگیرد (تا زمانی که کاربر `"exit"` تایپ کند) و همه خطوط را در فایلی به نام `user_input.txt` ذخیره کند.

### تمرین ۲ — خواندن فایل و شمارش خطوط
برنامه‌ای بنویسید که فایل `user_input.txt` را بخواند، تعداد خطوط، تعداد کلمات و تعداد کاراکترهای آن را چاپ کند.

### تمرین ۳ — اضافه کردن به فایل
برنامه‌ای بنویسید که یک خط جدید از کاربر بگیرد و آن را به همراه تاریخ و ساعت فعلی به انتهای فایل `user_input.txt` اضافه کند.

### تمرین ۴ — خواندن داده‌های ساده CSV
فرض کنید فایلی به نام `products.csv` با محتوای زیر دارید:

```csv
Name,Price,Quantity
Laptop,1200,5
Mouse,25,10
Keyboard,75,8
```

فایل را بخوانید، هر ردیف را پردازش کنید و مجموع ارزش موجودی انبار (`Price * Quantity`) را برای هر کالا و در کل چاپ کنید.

### تمرین ۵ — جستجو در فایل
برنامه‌ای بنویسید که یک کلمه از کاربر بگیرد، فایل `user_input.txt` را بخواند و تمام خطوطی که حاوی آن کلمه هستند را به همراه شماره خط چاپ کند.

---

## ۱۰.۱۳ خلاصه و نتیجه‌گیری

در این فصل با روش‌های مختلف کار با فایل در پایتون آشنا شدیم.

| موضوع | نکته کلیدی |
|---|---|
| **باز کردن فایل** | `open(filename, mode)` |
| **بهترین روش** | همیشه از `with open()` استفاده کنید |
| **خواندن** | `read()`, `readline()`, `readlines()`, حلقه `for` |
| **نوشتن** | `write()`, `writelines()` |
| **حالت‌ها** | `'r'` خواندن، `'w'` نوشتن، `'a'` اضافه کردن، `'b'` باینری |
| **انکدینگ** | همیشه `encoding="utf-8"` را مشخص کنید |
| **فایل‌های بزرگ** | خط به خط بخوانید، نه یکجا |
| **CSV** | از ماژول `csv` و `DictReader` استفاده کنید |

---

*فصل بعدی: مدیریت خطاها و استثناها در پایتون*
`
</div>
